torch/onnx/utils.py

```diff
@@ -46,48 +46,58 @@ def is_in_onnx_export():
+    else:
+        export_mode = mode
+
+    originally_training: bool = False
     if not isinstance(model, torch.jit.ScriptFunction):
-        is_originally_training = model.training
+        originally_training = model.training
 
         if mode is None:
-            mode = _C_onnx.TrainingMode.EVAL
             # if the model is in training mode but the user did not specify
             # to export the model in training mode, export the model in inference
             # mode (default) and warn them
-            if is_originally_training:
+            if originally_training:
                 warnings.warn(
                     "You are exporting the model to ONNX while in training mode with "
                     "'train' parameter not specified. The model will default to inference mode export. "
                     "If you wish to export a training amenable ONNX model, specify training=TrainingMode.TRAINING or "
                     "training=TrainingMode.PRESERVE (to preserve the original model state) in torch.onnx.export()."
                 )
 
-        # if mode == TrainingMode.EVAL or (mode == TrainingMode.PRESERVE and not is_originally_training) => is_training = False
-        is_export_training = False
         # ONNX opset 12 has better support for training amenable models, with updated
         # versions of the dropout and batch_norm operators
-        if mode == _C_onnx.TrainingMode.TRAINING or (
-            mode == _C_onnx.TrainingMode.PRESERVE and is_originally_training
+        if export_mode == _C_onnx.TrainingMode.TRAINING or (
+            export_mode == _C_onnx.TrainingMode.PRESERVE and originally_training
         ):
-
             if GLOBALS.export_onnx_opset_version < 12:
                 warnings.warn(
                     "You are exporting the model in training mode with onnx opset "
                     f"version {GLOBALS.export_onnx_opset_version}. "
                     "Opset versions lower than opset 12 will not be able to export "
                     "nodes such as Dropout and BatchNorm correctly."
                 )
-            is_export_training = True
 
-        symbolic_helper._set_training_mode(is_export_training)
-        model.train(is_export_training)
+        GLOBALS.training_mode = export_mode
+        if export_mode == _C_onnx.TrainingMode.TRAINING:
+            model.train(True)
+        elif export_mode == _C_onnx.TrainingMode.EVAL:
+            model.train(False)
+        # else mode == _C_onnx.TrainingMode.PRESERVE, do nothing
+
     try:
         yield
     finally:
-        if not isinstance(model, torch.jit.ScriptFunction):
-            # FIXME(justinchuby): is_originally_training is possibly unbound
-            model.train(is_originally_training)
+        if not (
+            isinstance(model, torch.jit.ScriptFunction)
+            or export_mode == _C_onnx.TrainingMode.PRESERVE
+        ):
+            # Alter the model training state only if export_mode is not PRESERVE
```

Comment:
nit: comment seems obvious and repeatetive to above code

Reply:
Done